flink-core/src/main/java/org/apache/flink/api/common/eventtime/WatermarksWithWatermarkAlignment.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.eventtime;
+
+import org.apache.flink.annotation.Internal;
+
+import java.time.Duration;
+
+/** A helper class to pass a watermark group and max allowed watermark drift to the runtime. */
+@Internal
+public final class WatermarksWithWatermarkAlignment<T> implements WatermarkStrategy<T> {
+
+    static final Duration DEFAULT_UPDATE_INTERVAL = Duration.ofMillis(1000);
+
+    private final WatermarkStrategy<T> strategy;
+
+    private final String watermarkGroup;
+
+    private final Duration maxAllowedWatermarkDrift;
+
+    private final Duration updateInterval;
+
+    public WatermarksWithWatermarkAlignment(
+            String watermarkGroup,
+            Duration maxAllowedWatermarkDrift,
+            Duration updateInterval,
+            WatermarkStrategy<T> strategy) {
+        this.strategy = strategy;
+        this.watermarkGroup = watermarkGroup;
+        this.maxAllowedWatermarkDrift = maxAllowedWatermarkDrift;
+        this.updateInterval = updateInterval;
+    }
+
+    public String getWatermarkGroup() {
+        return watermarkGroup;
+    }
+
+    public Duration getMaxAllowedWatermarkDrift() {
+        return maxAllowedWatermarkDrift;
+    }
+
+    public Duration getUpdateInterval() {
+        return updateInterval;
+    }
+
+    @Override
+    public WatermarkGenerator<T> createWatermarkGenerator(
+            WatermarkGeneratorSupplier.Context context) {
+        return strategy.createWatermarkGenerator(context);
+    }
+
+    @Override
+    public TimestampAssigner<T> createTimestampAssigner(TimestampAssignerSupplier.Context context) {
+        return strategy.createTimestampAssigner(context);
+    }
+
+    @Override
+    public WatermarkStrategy<T> withTimestampAssigner(
+            TimestampAssignerSupplier<T> timestampAssigner) {
+        return new WatermarksWithWatermarkAlignment<T>(
+                watermarkGroup,
+                maxAllowedWatermarkDrift,
+                updateInterval,
+                strategy.withTimestampAssigner(timestampAssigner));
+    }
```

Comment:
I'm not sure if such approach would scale well. I guess it would be working now, but:
1. if someone adds new method to the `WatermarkStrategy` interface with a default implementation, this will brake
2. if someone adds another strategy, that is implemented the same way as `WatermarksWithWatermarkAlignment` they will also not work together.

Instead, can not we keep the same pattern as before, so this class would only implement `createTimestampAssigner`, `createWatermarkGenerator` and a new method `getAlignmentParams()`? While replacing `watermarkStrategy instanceof WatermarksWithWatermarkAlignment` check with a `watermarkStrategy.getAlignmentParams()` call:
```
final class WatermarkStrategyWithAlignment<T> implements WatermarkStrategy<T> {

    private static final long serialVersionUID = 1L;

    private final WatermarkStrategy<T> baseStrategy;
    private final String watermarkGroup;
    private final Duration maxAllowedWatermarkDrift;
    private final Duration updateInterval;

    public WatermarkStrategyWithAlignment(
            String watermarkGroup,
            Duration maxAllowedWatermarkDrift,
            Duration updateInterval,
            WatermarkStrategy<T> strategy) {
        this.baseStrategy = strategy;
        this.watermarkGroup = watermarkGroup;
        this.maxAllowedWatermarkDrift = maxAllowedWatermarkDrift;
        this.updateInterval = updateInterval;
    }

    @Override
    public WatermarkAlignmentParams getAlignmentParams() {
        // default, return WatermarkAlignmentParams#DISABLED;
        return new WatermarkAlignmentParams(watermarkGroup, maxAllowedWatermarkDrift, updateInterval);
    }

    @Override
    public TimestampAssigner<T> createTimestampAssigner(TimestampAssignerSupplier.Context context) {
        return baseStrategy.createTimestampAssigner(context);
    }

    @Override
    public WatermarkGenerator<T> createWatermarkGenerator(
            WatermarkGeneratorSupplier.Context context) {
        ...
    }
}
```
?